content/blog/next-level-iac-briding-the-declarative-gap/index.md

```diff
@@ -0,0 +1,164 @@
+const bucket = new aws.s3.Bucket("bucket");
+
+// Export the name of the bucket
+export const bucketName = bucket.id;
+```
+
+{{% /choosable %}}
+
+{{% choosable language python %}}
+
+```python {hl_lines=[8]}
+import pulumi
+import pulumi_aws as aws
+
+# Create an S3 bucket
+bucket = aws.s3.Bucket('bucket')
+
+# Export the name of the bucket
+pulumi.export('bucket_name', bucket.id)
+```
+
+{{% /choosable %}}
+
+When this program runs, the call to create the `bucket` object will issue a request to the *deployment engine*. The deployment engine will load the *resource provider* for AWS, and ask it to create the S3 bucket. The AWS resource provider then issues the request to the AWS API, which will create the bucket and then return the `id` of the newly created resource. The provider will return that to the deployment engine, which will then *set the realized value* on your Pulumi program's `bucket.id` property and then move on to creating anything that depends on that property.
+
+This all seems pretty straight-forward, except that Pulumi programs don't wait and imperatively do each operation, one at a time. Instead the deployment engine will run them in parallel, only waiting in cases where the output of one operation is needed to run the next operation. This can create complex structures with some operations needing to wait for inputs from other operations, and some that don't, all finishing at different times, depending on how long the cloud service takes to process that request.
+
+## Inputs and Outputs as asynchronous futures
+
+Infrastructure automation tasks can be long-running and can result in intermittent errors and may need to be retried before moving on to the next step. A robust IaC solution will need to run concurrently, be able to restart, retry, and resume itself, and understand that a set of tasks could happen in any order, not necessarily in the order you defined them in code.
+
+Back in our Pulumi program code, that means that some of the time the `bucket` object from our earlier example has its `id` property set, and sometimes it doesn't. We wouldn't want to run any operation that needed that variable before it was set, but we also wouldn't want to wait to start operations that don't need that value.
+
+To address this, Pulumi models the input and output values as [futures][futures-wiki]. This allows you to describe a complex set of needs using a simple value-like syntax. Instead of using a `string` result directly, which would only exist some of the time, you instead work with an [`Output<string>`][output-docs], that Pulumi understands as a reference to a value that currently doesn’t exist, but *will* exist when the operation completes. Similarly, [`Input<string>`][input-docs] describes a dependency on an input string that will *at some point* be provided as the result of another operation. This allows Pulumi to execute the entirety of your graph of infrastructure tasks in parallel, only waiting when necessary.
```

Comment:
Can you elaborate here what you mean by "but will exist when the operation completes". I think there is a lot of hidden goodness in that phrase. You can probably talk about how the deployment agent waits on that value before completing the operation...

Reply:
I added a bit more explanation to cover this.